packages/aws-cdk/lib/notices.ts

```diff
@@ -4,76 +4,245 @@ import * as path from 'path';
+  readonly showTotal?: boolean;
+
+  /**
+   * Printer function.
+   *
+   * @default print
+   */
+  readonly printer?: (message: string) => void;
 
+}
+
+export interface NoticesForCommonOptions {
   /**
-   * Whether to append the number of unacknowledged notices to the display.
+   * Include only unacknowledged notices.
    *
    * @default false
    */
-  readonly unacknowledged?: boolean;
+  readonly onlyUnacknowledged?: boolean;
 }
 
-export async function refreshNotices() {
-  const dataSource = dataSourceReference(false);
-  return dataSource.fetch();
+export interface NoticesForCliVersionOptions extends NoticesForCommonOptions {
+  /**
+   * The version of the CLI to filter notices for.
+   *
+   * @default - The current version of the CLI
+   */
+  readonly cliVersion?: string;
 }
 
-export async function displayNotices(props: DisplayNoticesProps) {
-  const dataSource = dataSourceReference(props.ignoreCache ?? false);
-  print(await generateMessage(dataSource, props));
-  return 0;
+export interface NoticesForBootstrapVersionOptions extends NoticesForCommonOptions {
+  /**
+   * The version of Bootstrap to filter notices for.
+   */
+  readonly bootstrapVersion: number;
 }
 
-export async function generateMessage(dataSource: NoticeDataSource, props: DisplayNoticesProps) {
-  const data = await dataSource.fetch();
-  const filteredNotices = filterNotices(data, {
-    outdir: props.outdir,
-    acknowledgedIssueNumbers: new Set(props.acknowledgedIssueNumbers),
-  });
+export interface NoticesForFrameworkVersionOptions extends NoticesForCommonOptions {
+  /**
+   * The directory where the CDK application is synthesized to.
+   *
+   * @default - taken from the configuration or 'cdk.out'
+   */
+  readonly outDir?: string;
+}
+
+export interface NoticesRefreshOptions {
+  /**
+   * Whether to force a cache refresh regardless of expiration time.
+   *
+   * @default false
+   */
+  readonly force?: boolean;
+
+  /**
+   * Data source for fetch notices from.
+   *
+   * @default - CachedDataSource which fetches from our s3 bucket and populates a cache file.
+   */
+  readonly dataSource?: NoticeDataSource;
+}
 
-  let messageString: string = '';
-  if (filteredNotices.length > 0) {
-    messageString = getFilteredMessages(filteredNotices);
+/**
+ * Provides access to notices the CLI can display.
+ */
+export class Notices {
+
+  /**
+   * Create an instance. Note that this replaces the singleton.
+   */
+  public static create(props: NoticesProps): Notices {
+    this._instance = new Notices(props);
+    return this._instance;
   }
-  if (props.unacknowledged) {
-    messageString = [messageString, `There are ${filteredNotices.length} unacknowledged notice(s).`].join('\n\n');
+
+  /**
+   * Get the singleton instance. Note this will throw if `create` was not called.
+   */
+  public static get(): Notices {
+    if (!this._instance) {
+      throw new Error('Notices have not been initialized. Call create.');
```

Comment:
Interested in thoughts here. We could also return `undefined` here if this is too aggressive.

Reply:
I think this is too aggressive, because one of the tenets of this feature is that it should be totally non-intrusive. If this fails for whatever reason, it should not interfere with the actual command the user is running. But returning `undefined` is also disruptive. Ideally, we should return a no-op implementation in this case.